presto-hive/src/test/java/com/facebook/presto/hive/statistics/TestQuickStatsProvider.java

```diff
@@ -0,0 +1,341 @@
+        expectedPartitionStats = convertToPartitionStatistics(mockPartitionQuickStats);
+    }
+
+    @Test
+    public void testReadThruCaching()
+    {
+        QuickStatsBuilder quickStatsBuilderMock = mock(QuickStatsBuilder.class);
+        when(quickStatsBuilderMock.buildQuickStats(any(), any(), any(), any(), any(), any()))
+                .thenReturn(mockPartitionQuickStats);
+
+        QuickStatsProvider quickStatsProvider = new QuickStatsProvider(hdfsEnvironment, directoryListerMock, hiveClientConfig, new NamenodeStats(),
+                ImmutableList.of(quickStatsBuilderMock));
+
+        // Execute
+        ImmutableList<String> testPartitions1 = ImmutableList.of("partition1", "partition2", "partition3");
+        Map<String, PartitionStatistics> quickStats = quickStatsProvider.getQuickStats(SESSION, metastoreMock,
+                new SchemaTableName(TEST_SCHEMA, TEST_TABLE), metastoreContext, testPartitions1);
+
+        // Verify only one call was made for each test partition
+        assertEquals(quickStats.entrySet().size(), testPartitions1.size());
+        assertTrue(quickStats.keySet().containsAll(testPartitions1));
+        quickStats.values().forEach(ps -> assertEquals(ps, expectedPartitionStats));
+        for (String testPartition : testPartitions1) {
+            verify(quickStatsBuilderMock).buildQuickStats(any(), any(), any(), any(), eq(testPartition), any());
+        }
+        verifyNoMoreInteractions(quickStatsBuilderMock);
+
+        // For subsequent calls for the same partitions that are already cached, no new calls are mode to the quick stats builder
+        quickStatsProvider.getQuickStats(SESSION, metastoreMock,
+                new SchemaTableName(TEST_SCHEMA, TEST_TABLE), metastoreContext, testPartitions1);
+        verifyNoMoreInteractions(quickStatsBuilderMock);
+
+        // For subsequent calls with a mix of old and new partitions, we only see calls to the quick stats builder for the new partitions
+        ImmutableList<String> testPartitions2 = ImmutableList.of("partition4", "partition5", "partition6");
+        ImmutableList<String> testPartitionsMix = ImmutableList.<String>builder().addAll(testPartitions1).addAll(testPartitions2).build();
+
+        quickStats = quickStatsProvider.getQuickStats(SESSION, metastoreMock,
+                new SchemaTableName(TEST_SCHEMA, TEST_TABLE), metastoreContext, testPartitionsMix);
+        assertEquals(quickStats.entrySet().size(), testPartitionsMix.size());
+        assertTrue(quickStats.keySet().containsAll(testPartitionsMix));
+        quickStats.values().forEach(ps -> assertEquals(ps, expectedPartitionStats));
+        for (String testPartition : testPartitions2) {
+            verify(quickStatsBuilderMock).buildQuickStats(any(), any(), any(), any(), eq(testPartition), any());
+        }
+        verifyNoMoreInteractions(quickStatsBuilderMock);
+    }
+
+    @Test(invocationCount = 3)
```

Comment:
Why 3 invocations?

Reply:
To remove any possible noise from thread timing - `future2` may be the one to schedule the in-progress build and that shouldn't break the test